src/main/cues/cue-emitter.js

```diff
@@ -0,0 +1,224 @@
+  }
+
+  CueEmitter.prototype = Object.create(EventEmitter.prototype);
+  CueEmitter.prototype.constructor = CueEmitter;
+
+  CueEmitter.prototype._updateMarks = function() {
+    populateMarks(this.marks, this.peaks);
+  };
+
+  CueEmitter.prototype._onUpdate = function(time, previousTime) {
+    var marks = this.marks;
+    var isForward = time > previousTime;
+    var start;
+    var end;
+    var step;
+
+    if (isForward) {
+      start = 0;
+      end = marks.length;
+      step = 1;
+    }
+    else {
+      start = marks.length - 1;
+      end = -1;
+      step = -1;
+    }
+    // marks are sorted
+    for (var i = start, mark, markTime; isForward ? i < end : i > end; i += step) {
+      mark = marks[i];
+      markTime = mark.getTime();
+
+      if (isForward ? markTime > previousTime : markTime < previousTime) {
+        if (isForward ? markTime > time : markTime < time) {
+          break;
+        }
+        // mark time falls between now and previous call time
+        mark.emitEvent(this, isForward);
+      }
+    }
+  };
+
+  // the next handler and onAnimationFrame are bound together
+  // when the window isn't in focus, rAF is throttled
+  // falling back to timeUpdate
+
+  CueEmitter.prototype.onTimeUpdate = function(time) {
+    if (windowIsVisible()) {
+      return;
+    }
+    var el = getMediaElement(this.peaks);
+
+    if (this.peaks.player.isPlaying() && !el.seeking) {
+      this._onUpdate(time, this.previousTime);
+    }
+    this.previousTime = time;
+  };
+
+  CueEmitter.prototype.onAnimationFrame = function() {
+    if (!this.peaks) {
+      return;
+    } // destroyed
+
+    var peaks = this.peaks;
+    var el = getMediaElement(this.peaks);
+    var time = peaks.player.getCurrentTime();
+    var isPlaying = peaks.player.isPlaying();
+
+    if (!el.seeking) {
+      this._onUpdate(time, this.previousTime);
+    }
+    this.previousTime = time;
+    if (isPlaying) {
+      rAF(this.onAnimationFrame);
+    }
+  };
+
+  CueEmitter.prototype.onPlay = function() {
+    this.previousTime = this.peaks.player.getCurrentTime();
+    rAF(this.onAnimationFrame);
+  };
+
+  CueEmitter.prototype.onSeek = function() {
+    this.previousTime = this.peaks.player.getCurrentTime();
+  };
+
+  CueEmitter.prototype.onPeaksDestroyed = function() {
+    this.destroy();
+  };
+
+  var triggerUpdateOn = Array(
+    'points.update',
+    'points.dragmove',
+    'points.add',
+    'points.remove',
+    'points.remove_all',
+    'segments.update',
+    'segments.dragged',
+    'segments.add',
+    'segments.remove',
+    'segments.remove_all'
+  );
+
+  CueEmitter.prototype.attach = function() {
+    var peaks = this.peaks;
+
+    peaks.on('player_time_update', this.onTimeUpdate);
+    peaks.on('player_play', this.onPlay);
+    peaks.on('player_seek', this.onSeek);
+    peaks.on('destroyed', this.onPeaksDestroyed);
+    for (var i = 0; i < triggerUpdateOn.length; i++) {
+      peaks.on(triggerUpdateOn[i], this._updateMarks);
```

Comment:
These operations get more expensive with increasing numbers of points and segments.

Reply:
redoing the marks is triggered by a user action, it's outside the nominal (hot) loop when playing? I wanted to bind it to `points.update` and `segments.update` only but it seems the UI modifies them directly.

Reply:
> I wanted to bind it to points.update and segments.update only but it seems the UI modifies them directly.

This looks like an oversight, editing points/segments via the UI should probably emit `update` events.